ustream-ssl-wolfssl: add missing openssl dependency

mfrischknecht opened the pull request:
`ustream-ssl` still needs `openssl` as a build input if it is built against `wolfssl` (see [^1]). I haven't deciphered the build setup entirely, but from the respective CMake declarations, it looks like the library still won't be linked against `openssl` if it is built against `wolfssl` (see the `SET(SSL_LIB ...)` declarations), so I guess `openssl` is solely needed for some header definitions.

[^1]: https://git.openwrt.org/?p=project/ustream-ssl.git;a=blob;f=CMakeLists.txt;h=2de65905a40d8893155877a2b0894bcedec56f39;hb=HEAD#l24

## Description of changes

- Add `openssl` as a build input if `ustream-ssl` is built against `wolfssl`

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mfrischknecht commented:
Example of a Hydra job that is failing because of this issue: https://hydra.nixos.org/build/242943226
Hydra log: https://hydra.nixos.org/build/242943226/nixlog/4

Relevant log excerpt:

```
-- Build files have been written to: /build/ustream-ssl-498f6e2/build
cmake: enabled parallel building
cmake: enabled parallel installing
Running phase: buildPhase
build flags: -j2 SHELL=/nix/store/l34lhs5k16k8pydj2rjq70pqxz86kd3f-bash-5.2-p21/bin/bash
[ 25%] Building C object CMakeFiles/ustream-ssl.dir/ustream-io-openssl.c.o
[ 25%] Building C object CMakeFiles/ustream-ssl.dir/ustream-ssl.c.o
In file included from /build/ustream-ssl-498f6e2/ustream-io-openssl.c:24:
/build/ustream-ssl-498f6e2/openssl_bio_compat.h:4:10: fatal error: openssl/opensslv.h: No such file or directory
    4 | #include <openssl/opensslv.h>
      |          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/ustream-ssl.dir/build.make:90: CMakeFiles/ustream-ssl.dir/ustream-io-openssl.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /build/ustream-ssl-498f6e2/ustream-internal.h:27,
                 from /build/ustream-ssl-498f6e2/ustream-ssl.c:25:
/build/ustream-ssl-498f6e2/ustream-openssl.h:26:10: fatal error: openssl/ssl.h: No such file or directory
   26 | #include <openssl/ssl.h>
      |          ^~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/ustream-ssl.dir/build.make:76: CMakeFiles/ustream-ssl.dir/ustream-ssl.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:87: CMakeFiles/ustream-ssl.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

mfrischknecht commented:
@fpletz @mkg20001

I haven´t yet had to perform a similar modification to a derivation that is parameterized to such a degree. Do you think the `lib.hasPrefix "wolfssl" ssl_implementation.pname` check is adequate here in order to detect whether `ustream-ssl` is built against `wolfssl` or not? Is there maybe a better way to do this?

mkg20001 commented:
@mfrischknecht Another idea is making ssl_implementation a list and then using (head ssl_implementation).pname instead of ssl_implementation.pname and buildInputs = ssl_implementation

This would make it easily expandable. But I'm not sure if it fits the style of nixpkgs.

mfrischknecht committed:
ustream-ssl-wolfssl: add missing openssl dependency

`ustream-ssl` still needs `openssl` as a build input if it is built
against `wolfssl` (see [1]). I haven't deciphered the build setup
entirely, but from the respective CMake declarations, it looks like
the library still won't be linked against `openssl` if it is built
against `wolfssl` (see the `SET(SSL_LIB ...)` declarations), so I
guess `openssl` is solely needed for some header definitions.

[1]: https://git.openwrt.org/?p=project/ustream-ssl.git;a=blob;f=CMakeLists.txt;h=2de65905a40d8893155877a2b0894bcedec56f39;hb=HEAD#l24

mfrischknecht commented:
@mkg20001

> @mfrischknecht Another idea is making ssl_implementation a list and then using (head ssl_implementation).pname instead of ssl_implementation.pname and buildInputs = ssl_implementation
> 
> This would make it easily expandable. But I'm not sure if it fits the style of nixpkgs.

Thank you for your help :)

I actually tried something like this, but the `preConfigure` phase actually contains [logic that would stop working without a similarly complex refactoring](https://github.com/NixOS/nixpkgs/blob/57cea6b6fdde7258a9610dc41f229d9388b002fb/pkgs/development/libraries/ustream-ssl/default.nix#L17) if `ssl_implementation` were a list of packages:

```nix
  preConfigure = ''
    sed -r \
# ...
        -e "s|^  FIND_LIBRARY\((.+)\)|  FIND_LIBRARY\(\1 HINTS ${if ssl_implementation ? lib then ssl_implementation.lib else ssl_implementation.out}\)|g" \
        -i CMakeLists.txt
  '';
```

But following your thought, I could imagine introducing e.g. an argument like `additional_buildInputs ? [ ]` to the function and then passing `openssl` in from the callsite like this (see https://github.com/mfrischknecht/nixpkgs/commit/982c21722a3aa69b18f01252642425caeb008777 for a working example):

```nix
ustream-ssl-wolfssl = callPackage ../development/libraries/ustream-ssl { ssl_implementation = wolfssl; additional_buildInputs = [ openssl ]; };
```

https://github.com/mfrischknecht/nixpkgs/commit/982c21722a3aa69b18f01252642425caeb008777 also shows that with this approach, the change is indeed much simpler -- but at the same time, this goes somewhat against my instincts on information hiding; my gut tells me that `ustream-ssl` needing `openssl` when built against `wolfssl` is an implementation detail a larger scope shouldn't really be concerned about, and adding the argument basically distributes this logic over more files.

But then again, I don't know what an idiomatic Nixpkgs solution would look like ¯\_(ツ)_/¯.

What do you think, should I switch the PR to the version with the additional parameter?

mkg20001 commented:
> my gut tells me that ustream-ssl needing openssl when built against wolfssl is an implementation detail a larger scope shouldn't really be concerned about, and adding the argument basically distributes this logic over more files.

we already put that logic into another file (with ssl_implementation = wolfssl; instead of using "useWolfssl" as a flag on the package) so going with the additional argument sounds like the logical thing to do

mfrischknecht commented:
> we already put that logic into another file (with ssl_implementation = wolfssl; instead of using "useWolfssl" as a flag on the package) so going with the additional argument sounds like the logical thing to do

Okay, thanks :) I've changed the PR branch accordingly.
